source/hp/fe_collection.cc

```diff
@@ -103,6 +109,69 @@ namespace hp
 
 
 
+  template <int dim, int spacedim>
+  unsigned int
+  FECollection<dim, spacedim>::find_most_face_dominating_fe_in_subset(
+    const std::set<unsigned int> &fes) const
+  {
+    const hp::FECollection<dim, spacedim> &fe_collection = *this;
+
+    // If the set of elements to be dominated contains only a single element X,
+    // then by definition the dominating set contains this single element
+    // (because each element can dominate itself).
+    // There may also be others, in which case we'll check if any of these
+    // elements is able to dominate all others. If one was found, we stop
+    // looking further and return the dominating element.
+    if (fes.size() == 1)
+      {
+        Assert(*fes.begin() < fe_collection.size(),
+               ExcIndexRangeType<unsigned int>(*fes.begin(),
+                                               0,
+                                               fe_collection.size()));
+
+        return *fes.begin();
+      }
+
+    // first loop over all FEs and check which can dominate those given in @p fes:
+    for (const auto &current_fe : fes)
+      {
+        Assert(current_fe < fe_collection.size(),
+               ExcIndexRangeType<unsigned int>(current_fe,
+                                               0,
+                                               fe_collection.size()));
+
+        FiniteElementDomination::Domination domination =
+          FiniteElementDomination::either_element_can_dominate;
+
+        for (const auto &other_fe : fes)
+          if (other_fe != current_fe)
+            {
+              Assert(other_fe < fe_collection.size(),
+                     ExcIndexRangeType<unsigned int>(other_fe,
```

Comment:
this is checked quadratically often here because it's inside the nested loop. Just put the check into a separate loop at the beginning of the function. This then also covers the case above where we only have a single element.